src/box/gc.c

```diff
@@ -919,8 +922,22 @@ gc_erase_consumer(const struct tt_uuid *uuid)
 	struct trigger *on_commit =
 		xregion_alloc_object(&in_txn()->region, struct trigger);
 	trigger_create(on_commit, gc_erase_consumer_on_commit, consumer, NULL);
-	txn_stmt_on_commit(txn_current_stmt(in_txn()), on_commit);
-	return 0;
+	if (is_autocommit) {
+		/*
+		 * In the case of autocommit we can set the trigger right to the
+		 * txn since we will commit it and the statement won't be rolled
+		 * back.
+		 */
+		txn_on_commit(in_txn(), on_commit);
+	} else {
+		/*
+		 * If external txn is used, the statement can be rolled back
+		 * before its commit, so we should set the trigger to the
+		 * current statement, not the txn itself.
+		 */
+		txn_stmt_on_commit(txn_current_stmt(in_txn()), on_commit);
```

Comment:
Why not simply set it on the statement always?

Reply:
Because in auto commit mode we have no "current statement" - a statement that is being processed right now. In other cases we have it because we call the function only on `_cluster:replace`, so the replace in `_cluster` is actually a current statement.

But after your question I realised that it looks weird, so I introduced a new helper `txn_last_stmt` and use it here to always put the `on_commit` trigger to the actual `_gc_consumers:delete` statement.